java/vector/src/main/java/org/apache/arrow/vector/BaseLargeVariableWidthVector.java

```diff
@@ -336,6 +336,31 @@ public List<ArrowBuf> getFieldBuffers() {
     return result;
   }
 
+  /**
+   * Export the buffers of the fields for C Data Interface. This method traverse the buffers and
+   * export buffer and buffer's memory address into a list of buffers and a pointer to the list of buffers.
+   */
+  @Override
+  public void exportCDataBuffers(List<ArrowBuf> buffers, ArrowBuf buffersPtr, long nullValue) {
+    // before flight/IPC, we must bring the vector to a consistent state.
+    // this is because, it is possible that the offset buffers of some trailing values
+    // are not updated. this may cause some data in the data buffer being lost.
+    // for details, please see TestValueVector#testUnloadVariableWidthVector.
+    fillHoles(valueCount);
+
+    exportBuffer(validityBuffer, buffers, buffersPtr, nullValue, true);
+
+    if (offsetBuffer.capacity() == 0) {
+      // Empty offset buffer is allowed for historical reason.
+      // To export it through C Data interface, we need to allocate a buffer with one offset.
+      exportBuffer(allocateOffsetBuffer(OFFSET_WIDTH), buffers, buffersPtr, nullValue, false);
```

Comment:
@viirya I just want to make sure I understand what is happening here. It seems we are creating a new buffer which will have ref count of zero, then you specify `retain = false` when calling `exportBuffer` so the ref count is still zero, right? Then the export will increment to 1, which is the desired behavior.

Reply:
Yes. By default when creating a new buffer the allocator will increase ref count to retain the buffer for the holder (i.e., the vector). The array exporter increases ref count that represents the buffer is retained by the importer.

This offset buffer is not retained by the vector actually, but only for the importer. So it doesn't increase the ref count when creating it (as it is not held by the vector). In array exporter, it will increase ref count for the buffer as usual.

Reply:
By default when creating a new buffer the allocator will increase ref count to retain the buffer for the holder, so its initial ref count is 1.

When we are going to export an empty offset buffer, we create a new offset with one offset value zero. `exportBuffer` is called with `retain = false` so it won't increase its ref count. Its ref count is 1 which is for importer.

For other buffers, `exportBuffer` will increase ref count by 1 which represents a new reference by the importer.

Reply:
Added some more comments.

Reply:
Thanks for the clarification @viirya. 

@lidavidm Does this address the specific concern that you had about how reference counting is working here?

Reply:
Probably, thanks for the explanation. I still want to take a closer look which I'll try to do in time for 16.0.0 code freeze but I will probably need a few days before I can get to this.